src/firmware/main/app/primitives/primitive_manager.c

```diff
@@ -0,0 +1,187 @@
+#include "firmware/main/app/primitives/pivot.h"
+#include "firmware/main/app/primitives/primitive.h"
+#include "firmware/main/app/primitives/shoot.h"
+#include "firmware/main/app/primitives/spin.h"
+#include "firmware/main/app/primitives/stop.h"
+
+struct PrimitiveManager
+{
+// The mutex that prevents multiple entries into the same primitive at the same time.
+#ifdef __arm__
+    SemaphoreHandle_t primitive_mutex;
+#elif __unix__
+    pthread_mutex_t primitive_mutex;
+#endif
+
+    // TODO: rename, this is really a pointer to the _functions_ of the current
+    //       primitive, we hold it's state
+    // The primitive that is currently operating.
+    const primitive_t *current_primitive;
+
+    // The index number of the current primitive.
+    unsigned int current_primitive_index;
+
+    // A pointer to the state of the current primitive
+    void *current_primitive_state;
+};
+
+/**
+ * \brief The available movement primitives.
+ *
+ * This array is indexed by movement primitive number. If movement primitives
+ * are added or removed, they must be added or removed in this array and it
+ * must be kept in the same order as the enumeration in @c
+ * software/mrf/constants.h.
+ *
+ * Make sure stop is always the first primitive.
+ */
+static const primitive_t *const PRIMITIVES[] = {
+    &STOP_PRIMITIVE,             // index 0, do not change the order of stuff in this one
+    &MOVE_PRIMITIVE,             // 1
+    &DRIBBLE_PRIMITIVE,          // 2
+    &SHOOT_PRIMITIVE,            // 3
+    &CATCH_PRIMITIVE,            // 4
+    &PIVOT_PRIMITIVE,            // 5
+    &SPIN_PRIMITIVE,             // 6
+    &DIRECT_WHEELS_PRIMITIVE,    // 7
+    &DIRECT_VELOCITY_PRIMITIVE,  // 8
+    &IMU_TEST_PRIMITIVE,         // 9
+};
+
+/**
+ * \brief The number of primitives.
+ */
+#define PRIMITIVE_COUNT (sizeof(PRIMITIVES) / sizeof(*PRIMITIVES))
+
+/**
+ * Lock the primitive mutex
+ * @param manager The primitive manager to lock the primitive mutex for
+ */
+void app_primitive_manager_lockPrimitiveMutex(PrimitiveManager_t *manager)
+{
+#ifdef __arm__
+    xSemaphoreTake(manager->primitive_mutex, portMAX_DELAY);
+#elif __unix__
+    pthread_mutex_lock(&(manager->primitive_mutex));
+#else
+#error "Could not determine what CPU this is being compiled for."
+#endif
+}
+
+/**
+ * Unlock the primitive mutex
+ * @param manager The primitive manager to unlock the primitive mutex for
+ */
+void app_primitive_manager_unlockPrimitiveMutex(PrimitiveManager_t *manager)
+{
+#ifdef __arm__
+    xSemaphoreGive(manager->primitive_mutex);
+#elif __unix__
+    pthread_mutex_unlock(&(manager->primitive_mutex));
+#else
+#error "Could not determine what CPU this is being compiled for."
+#endif
+}
+
+PrimitiveManager_t *app_primitive_manager_create(void)
+{
+    PrimitiveManager_t *manager =
+        (PrimitiveManager_t *)malloc(sizeof(PrimitiveManager_t));
+
+#ifdef __arm__
+    // TODO: this could cause issues if we ever have multiple PrimitiveManager's running
```

Comment:
TODO?

Reply:
Also I don't think this is a concern we already made tons of assumptions about only one instance of stuff existing at once in firmware since it's per-robot

Reply:
Yeah, just wanted to come back and look at this, but I agree for now.